File: app/src/main/java/forpdateam/ru/forpda/model/repository/AuthRepository.java
package forpdateam.ru.forpda.model.repository;

import forpdateam.ru.forpda.api.auth.Auth;
import forpdateam.ru.forpda.api.auth.models.AuthForm;
import forpdateam.ru.forpda.api.favorites.Favorites;
import forpdateam.ru.forpda.model.system.SchedulersProvider;
import io.reactivex.Observable;

/**
 * Created by radiationx on 02.01.18.
 */

public class AuthRepository {

    private SchedulersProvider schedulers;
    private Auth authApi;

    public AuthRepository(SchedulersProvider schedulers, Auth authApi) {
        this.schedulers = schedulers;
        this.authApi = authApi;
    }

    public Observable<AuthForm> loadForm() {
        return Observable.fromCallable(() -> authApi.getForm())
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.ui());
    }

    public Observable<Boolean> signIn(AuthForm authForm) {
        return Observable.fromCallable(() -> authApi.login(authForm))
                .subscribeOn(schedulers.io())
                .observeOn(schedulers.ui());
    }

}

File: app/src/main/java/forpdateam/ru/forpda/model/repository/auth/AuthRepository.kt
package forpdateam.ru.forpda.model.repository.auth

import forpdateam.ru.forpda.entity.common.MessageCounters
import forpdateam.ru.forpda.entity.remote.auth.AuthCaptcha
import forpdateam.ru.forpda.entity.remote.auth.AuthForm
import forpdateam.ru.forpda.model.CountersHolder
import forpdateam.ru.forpda.model.data.remote.api.auth.AuthApi

/**
 * Created by radiationx on 02.01.18.
 */

class AuthRepository(
    private val authApi: AuthApi,
    private val countersHolder: CountersHolder,
) {

    suspend fun loadCaptcha(): AuthCaptcha {
        return authApi.getCaptcha()
    }

    suspend fun signIn(captcha: AuthCaptcha, form: AuthForm) {
        authApi.login(captcha, form)
    }

    suspend fun signOut() {
        authApi.logout()
        countersHolder.set(
            MessageCounters(
                mentions = 0,
                favorites = 0,
                qms = 0
            )
        )
    }

}
